### DISCUSSION TITLE
Converting T41 Breakout Board to support spindle sync

### DISCUSSION THREAD

**Discussion by phil-barrett** [terjeio/grblHAL#289] [General]
**Overview**

The Teensy 4.1 Breakout boards need a modification to support the recently developed spindle sync support in grblHAL.  Here are the instructions for modifying V2.07 and V2.09.  This covers most of the boards out there. If you have a different version and wish to make a change, let me know. Note that the Pro version (T41ProBB) does not need modification.

These instructions expect that you have skill in soldering, surface mount technology and working precisely with microelectronics.  I assume no liability for any damage that ensues from the use or misuse of these instructions. Not also that it will not be possible to undo these changes if you desire to revert to the original configuration of the board.

There are 3 general steps:

1. Move ST3 to Pin 14.
2. Move Halt to Pin 35
3. Modify the T41U5XBB.map file and rebuild grblHAL

**Materials and Tools**

You will need 

- a sharp knife with a pointed blade
- a soldering iron with a fine tip
- fine needle nose pliers
- a multimeter for continuity testing 
- 26 gauge insulated solid core wire. 

**1) ST3 to Pin 14**

Start on the top of the breakout board. You will need to cut one trace and add a bodge wire.  This requires care - there is little margin for error. Cut the trace indicated in the photo for your version. Be careful make sure you can cut through completely and there are no stray copper bits.  Also take care not to cut or damage adjacent wires or components.  Verify that you have no connectivity between pin 35 of the Teensy (labeled on the Teensy) and the bottom left pin of the IC as shown in the photo. If you still have continuity, your cut was not successful, revisit it.  

Now, measure a length of small wire (26 ga solid is ideal) for bodge 1. Strip a short length of insulation off one end and a bit longer off the other. Do not use uninsulated wire. Tin the short end with solder.  Carefully solder it to the bottom left pin of IC14 as shown in the photo. (Using needle nose pliers, hold the tinned end to the pin and place the soldering iron tip on it. When it melts remove the iron but hold the wire in place until the solder hardens.) Then solder the other end to the leftmost pin (labeled Hlt) on the 6 pin I2C header as indicated in the photo. Take care that no bare wire can contact any other pins, vias or solder pads. Test continuity between the IC pin you soldered the wire to and pin 14 of the Teensy (labeled on the Teensy). If you do not have continuity, recheck your soldering.  Also verify again that you do not have continuity between the IC14 pin and pin 35 of the Teensy.  Secure the bodge wire with some hot glue.

V2.07
![V 2 07 Cut 1 Bodge 1](https://user-images.githubusercontent.com/16570958/114602917-fb728a00-9c4b-11eb-8a5f-f55b7919da3e.jpg)

V2.09
![V 2 09 Cut 1 Bodge 1](https://user-images.githubusercontent.com/16570958/114602931-ff061100-9c4b-11eb-8871-a67e9fc1e105.jpg)

**Halt to Pin 35**

Turn the breakout board over and locate the area in the photos below. Cut the trace indicated in the photo. There should be no continuity between the pin directly above Cut 2 and the via where Bodge 2 starts.  Make up a bodge wire as in the first step including stripping and tinning. Solder one end to the via and the other end to pin 35 of the Teensy (third pin from the left as indicated in the photo.  Take care that no bare wire can contact other vias, pins or solder pads. Secure the bodge wire with hot glue.

V2.07
![V2 07 Cut 2 Bodge 2](https://user-images.githubusercontent.com/16570958/114605191-abe18d80-9c4e-11eb-8e75-a05462e29403.jpg)

V2.09
![V2 09 Cut 2 Bodge 2](https://user-images.githubusercontent.com/16570958/114605180-ab48f700-9c4e-11eb-94a1-aedcb5d950ca.jpg)

**Modify map file, rebuild**

Now, modify the T41U5XBB_map.h map file to tell grblHAL the new pin numbers. Starting at about line 109 you will find
```
// Define user-control CONTROLs (cycle start, reset, feed hold, door) input pins.
#define RESET_PIN           (14u)
#define FEED_HOLD_PIN       (16u)
#define CYCLE_START_PIN     (17u)
#define SAFETY_DOOR_PIN     (29u)

// Define probe switch input pin.
#define PROBE_PIN           (15U)

// Define auxillary input pins
#define AUXINPUT0_PIN       (36u) // ST0
#if !QEI_ENABLE
#define AUXINPUT1_PIN       (30u) // ST1
#define AUXINPUT2_PIN       (34u) // ST2
#define AUXINPUT3_PIN       (35u) // ST3

```
Swap 14u and 35u
```
// Define user-control CONTROLs (cycle start, reset, feed hold, door) input pins.
#define RESET_PIN           (35u)
#define FEED_HOLD_PIN       (16u)
#define CYCLE_START_PIN     (17u)
#define SAFETY_DOOR_PIN     (29u)

// Define probe switch input pin.
#define PROBE_PIN           (15U)

// Define auxillary input pins
#define AUXINPUT0_PIN       (36u) // ST0
#if !QEI_ENABLE
#define AUXINPUT1_PIN       (30u) // ST1
#define AUXINPUT2_PIN       (34u) // ST2
#define AUXINPUT3_PIN       (14u) // ST3

```
Rebuild and reload your new firmware.  You are ready to test.  [edit] what are the specific steps to add the plug-in to grblHAL, I am a little unclear on that[/edit]

**Comment by terjeio** [comment]
> Rebuild and reload your new firmware. You are ready to test. [edit] what are the specific steps to add the plug-in to grblHAL, I am a little unclear on that[/edit]

I'll create and upload a new map file for T41U5XBB with spindle sync definitions tomorrow. There is no plug-in to add, the spindle sync code is part of _driver.c_ and is enabled via _#define_ symbols.

**Reply by phil-barrett** [reply]
Perfect!  Thanks.

**Reply by terjeio** [reply]
The [new repo](https://github.com/grblHAL/iMXRT1062) has been updated for this modification.

Connect main pulse input to ST3 and index pulse to ST2, select `BOARD_T41U5XBB_SS` and uncomment `SPINDLE_SYNC_ENABLE` in _my_machine.h_.

Then configure spindle pulses per revolution (PPR) with `$38` and reboot the controller.
If using [ioSender](https://github.com/terjeio/Grbl-GCode-Sender) (recommended) then the RPM readout should now go live and display the real spindle RPM.

When this is working correctly proceed carefully to check if spindle sync is working by setting some initial PID loop parameters, e.g:

![image](https://user-images.githubusercontent.com/20260062/114706363-25e64500-9d29-11eb-83c8-af4e861dbcb5.png)

First step is to run some in-the air cuts to see if the index pulse is detected correctly, use a `G33` command for that.

If using ioSender then PID logging can be enabled in grbl/config.c by uncommenting

`#define PID_LOG 1000 // Default disabled. Uncomment to enable.`

at or near line 470 and reflash. This will make the _PID Tuner_ tab appear in the sender where the data from the last synced motion can be fetched and displayed. Very useful for tuning the PID loop parameters... I guess I'll have to come back to that.

Next step is to verify that multiple passes of synced motions overlap, best way to do that is to color a rod with a marker pen and run multiple very shallow cuts with `G33`. Check the video [here](https://hackaday.io/project/165248-mini-lathe-emco-compact-5-cnc-conversion) for how I did it.

grblHAL has a lathe mode that can be enabled with `$32=2`, when enabled ioSender will change the UI to reflect that. Among the changes are new tabs for conversational gcode generation for some lathe operations.

**WARNING!** I have not tested the spindle sync code for the iMXRT1062 driver in a machine and the code generated by ioSender has not seen much testing. Use with extreme care, lathes are dangerous!

**Reply by timgow** [reply]
Phil, terjeio,
Thank you for producing this fantastic guide to follow; I'm currently digesting.
I'll get on with building a v2.07 board and hope to be testing out the spindle sync soonest.

**Comment by NeilMarley** [comment]
This is perfect timing... I've just got my T41U5XBB running my Emco Compact 5 PC nicely and was about to ask how to activate spindle sync :-)
It's not a great issue to do the mods but can I ask if it is necessary to remap the spindle index & count signals as detailed here - is it not possible just to remap e.g. some of the unused limit-switch inputs and use those or do we need specific pins to feed the interrupt timers for spindle-sync?

Neil

**Reply by terjeio** [reply]
> is it not possible just to remap e.g. some of the unused limit-switch inputs and use those or do we need specific pins to feed the interrupt timers for spindle-sync?

It might be possible but will require a deep dive into the datasheet and possibly swapping timers around. I am not too keen to do that now as it is likely to complicate the driver code now that the Pro board is completed.

**Reply by phil-barrett** [reply]
It is about specific pins.  The Limit switches use Opto Isolators that are probably too slow for that application, any way. The Pro Teensy board does not need any modifications.

**Comment by NeilMarley** [comment]
That’s fine, as expected. 

Thanks for the prompt response guys - I’ll get the soldering iron out (was already deployed as I was boxing-up the controller 😊

 

I’m only driving 2 axes up to 24mm/s at the moment so nothing is being stressed….

It might be worth considering that many of the target audience are focussing on 2-3 axes mechanical lathe/mills and not using 5 axis flood-cooled behemoths? 

 

Neil

**Reply by NeilMarley** [reply]
No problem, mod(s) done as per excellent instructions. I've sent Phil some showcase pics....
All functional - some issues but will report once I've done a deep(er) dive diagnosis.

Neil

**Comment by Team-Deepflights** [comment]
I made the modification as described, I only have a normal encoder (6000 impulse pm) for testing, but it only counts to 365 rpm and then the display goto zero. The encoder was connected to ST3 only, it have no index which is connected to ST2.
Shouldn't it show at least the RPM or is the index signal required to reset the impulse counter?
Also I have a PID index tab but it shows nothing. same with the PID loop parameters, can't find them.
If i connect the board with a terminal and send "$$" I receive a lot of points, but no $90 to 95.

My lathe is a MX180D, this one have a photosensor to display the actual RPM. I try this one with the board (with 22 impulse per round) without success, RPM was shown as zero?

What I miss most in this instruction is an few examples what encoders could be used and the notice that the enconder send a Z Signal for threathing.

Any help is most appreciated, I need the threading option, this was the main reason to convert the lathe into CNC.

Merry X_mas btw :-)

**Reply by terjeio** [reply]
> ... I only have a normal encoder (6000 impulse pm) ...

Is that 6000 pulses per revolution (PPR)? Have you set $38 to match? I may have to add more prescaling if so (to reduce interrupt load).

> Shouldn't it show at least the RPM or is the index signal required to reset the impulse counter?

The index pulse is not needed for determining RPM, however it is required for spindle sync.

> Also I have a PID index tab but it shows nothing. same with the PID loop parameters, can't find them.

That is due to me commenting out code in _grbl/settings.c_, I have attached a fixed version below.

> My lathe is a MX180D, this one have a photosensor to display the actual RPM. I try this one with the board (with 22 impulse per round) without success, RPM was shown as zero?

That is odd, was $38 set to match the PPR? I'll tune my simulator to see if I can reproduce that.

> What I miss most in this instruction is an few examples what encoders could be used and the notice that the enconder send a Z Signal for threathing.

In principle any encoder that delivers an index signal once per revolution and a number of pulses per revolution can be used providing that the signal levels match what the controller can handle. The encoder does not send a Z signal, it is used to calculate the actual angular position every 5 ms. The actual position then is compared with the programmed position and the delta is used to adjust the Z step rate via a PID control loop.

> Any help is most appreciated, I need the threading option, this was the main reason to convert the lathe into CNC.

I'll try my best, not many has tinkered with spindle sync - and I have been to busy with coding to do so myself. Hopefully I will be back in the workshop soon - I want to complete my lathe conversion.

[settings.zip](https://github.com/terjeio/grblHAL/files/7776807/settings.zip)

Note that after changing the PID parameters the controller has to be hard reset. A $38 change will also make the PID parameters "take". To be fixed in the next commit. Start with P gain set to 1 and I gain set to 0.01.


Merry Christmas.

**Comment by Team-Deepflights** [comment]
> Is that 6000 pulses per revolution (PPR)? Have you set $38 to match? I may have to add more prescaling if so (to reduce interrupt load).

$38 is set, I also have played around with the value. The conversion is running perfectly with the steppers, so I think it's something software-sided.
I have ordered another encoder now with 200 PPR and Z-signal, I hope this one is going to work.
It would have been perfect if I could have use the built in photosensor, 3D print another wheel with just one hole and use an additional photosensor or a hall sensor for spindle sync.
In that case I would have both, the RPM in the built in lathe display and the RPM signal passed to the board with one sensor.

Yesterday I tried external another photosensor, the $38 was set to 1 and I passed a piece of paper for interrrupting, the RPM rises to 5-7 for about 2 secs then stopped and show zero.
Different to the 6000 PPR encoder, which start counting after running over and return to a zero value, the RPM display did not show any value again. Only a hard reboot show up a RPM value again, but, as mentioned, only for a very short time.
This really confuses me, could this be an interrupt overflow or similar?
I connected the sensor to  the 3,3v pin and also to 5v, same result. 

> In principle any encoder that delivers an index signal once per revolution and a number of pulses per revolution can be used providing that the signal levels match what the controller can handle. The encoder does not send a Z signal, it is used to calculate the actual angular position every 5 ms. The actual position then is compared with the programmed position and the delta is used to adjust the Z step rate via a PID control loop.

Fully understood that, after a while ;-)
I also know that you do this excellent work in your leisure time, I just wanted to point to that.
When I bought the 2 boards from Phil I asked about the syncing and he send me a link to this thread, I saw that this is not a great thing and ordered profusely an encoder in Ebay, now I know that 6000 PPR is maibe too much and as I want to sync the spindle, an indexing signal (no matter if its a hall sensor or something else) is mandatory.

Maibe you find some time and insert this in the conversion manual, maibe it help others and they don't mess like me buying an useless encoder  :-)

**Reply by NeilMarley** [reply]
Its also worth noting that the ST1-3 inputs on the Teensie breakout board have Low-Pass filters on them at ~16KHz to reduce noise. With your 6000ppr encoder (i.e. 100 pulses/sec per RPM) the pulses would start to be filtered-out at around 160rpm which might explain it stopping to register at all by 360rpm.

A simple Index pulse can be done with a magnet attached to the spindle and a hall-effect detector - as used by units like this:
https://www.amazon.co.uk/DIGITEN-Digital-Tachometer-Proximity-Switch/dp/B00VKAT8A2/ref=asc_df_B00VKAT8A2/?tag=googshopuk-21&linkCode=df0&hvadid=310879398178&hvpos=&hvnetw=g&hvrand=11527575704683018000&hvpone=&hvptwo=&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=1006886&hvtargid=pla-569951226433&psc=1

**Reply by terjeio** [reply]
> With your 6000ppr encoder (i.e. 100 pulses/sec per RPM) the pulses would start to be filtered-out at around 160rpm which might explain it stopping to register at all by 360rpm.

Good catch, I can confirm this but the cutoff is a bit higher for me as I output a 50% duty cycle signal from my simulator. If the duty cycle is different from 50% I guess the cutoff can be significatly lower as the switch off time is delayed quite a bit more than the switch on time. A cure could be to remove C5 from the input filter?

![Untitled](https://user-images.githubusercontent.com/20260062/147414672-5d3fbae1-fbb5-4d72-b3ee-6b5eefe1511a.png)

**Comment by Team-Deepflights** [comment]
Thinking about it, I am not sure if a such a high resolution encoder is really needed...
But why it did not work by passing a sheet of paper with 1 PPR set in the menu?
I'm going to try this and remove C5 for a test after sending this posting. I'll tell you later what happened.

> A simple Index pulse can be done with a magnet attached to the spindle and a hall-effect detector - as used by units like this:

I have a few of them here,  you are right, it's way easier to generate the index signal with those sensors.
Mine are inductive, i just need to pass a piece of metal.
Thanks a lot for this.

**Reply by terjeio** [reply]
> Thinking about it, I am not sure if a such a high resolution encoder is really needed...

Definitely not, I have a 120 PPR encoder and that seems to work well. More than 1024 PPR is a waste of money? 
BTW - a prescaler is used to reduce interrupt load, configured from the PPR setting.

>But why it did not work by passing a sheet of paper with 1 PPR set in the menu?

It is due to this test?
    // If no (4) spindle pulses during last 250 ms assume RPM is 0

**Comment by Team-Deepflights** [comment]
> I'm going to try this and remove C5 for a test after sending this posting. I'll tell you later what happened.

This did it, after removing C5 with hot air, now the 6000 PPR encoder seems to work, NeilMarley pointed into the right direction :-)
But i am not going to use this encoder at all, its too much precision for the job it have to do. I'll put back the C into its place.

> It is due to this test?
// If no (4) spindle pulses during last 250 ms assume RPM is 0

This explains everything! I did not dig so deep into the code.
But this affect only the ST2, right?

Ok, I'll keep working and report back to you as soon as possible.

**Reply by NeilMarley** [reply]
Glad that helped.
Even so, a 6000ppr encoder is complete overkill. TerjeIo is using 120 ppr, mine is the 100ppr standard on an Emco5 PC (& is quite noisy at that) & threading seems perfectly good (I must do a write-up on the PID tuning :-).
Trying to mop-up 6000ppr with software/prescaling is just going to create unnecessary complication/processor load?

Neil

**Comment by Team-Deepflights** [comment]
Absolutely, there is no need for such a high resolution.
Well, I will archive it again, I am sure one day I need such an encoder for something....

**Comment by Team-Deepflights** [comment]
Just tested the built in photosensor, the wheel have 16 holes, so PPR is 16, I changed $38 value to this.
Nothing happens.

> It is due to this test?
// If no (4) spindle pulses during last 250 ms assume RPM is 0

I raised the RPM but still the IOSender wrote zero.
It begins to count when I put 65 as $38 value, of course the RPM is wrong.

I found this remark in drive.c, but I am not sure if 

> static spindle_encoder_t spindle_encoder = {
    .tics_per_irq = 4
};

Are tics_per_irq = 4 the mentioned spindle pulses in the remark?

For now it seems that any encoder signal with is <64 PPR will not work as intended.

**Comment by karoria** [comment]
Hi @terjeio @phil-barrett !
Thanks for all your hard work. I am bit confused on, does the spindle sync mean I can do rigid tapping in cnc mill? Or it is just for lathe purpose?
Regards,
Ravi

**Reply by terjeio** [reply]
> does the spindle sync mean I can do rigid tapping in cnc mill? 

No, I do not have a mill that can be used for testing support for rigid tapping so I am not able to do it. I guess a spindle with enough torque at low speeds is needed and maybe also control of acceleration/deceleration of spindle RPM during reversal. And a lot of taps?

**Reply by karoria** [reply]
OK. that means theoretically it is implemented. Its practical aspects still to be considered and code to be modified accordingly. I was thinking to use AC servo spindle (just like Vertical Machining Centers) in one of my machines with ATC.
Another quick question, can I use pulse-direction interface for spindle? (I want to use AC servo motor for spindle). If that is possible, all the encoder feedback will be taken care by its drive and I might be able to do rigid tapping. I am sure you are not in a position to answer practical aspects of it but just a wild idea from me for your comments please.

**Reply by terjeio** [reply]
> Another quick question, can I use pulse-direction interface for spindle?

Yes, if you write code for it. Your code (either implementing _my_plugin_init()_ or _board_init()_) can "claim" spindle control. The Huanyang [spindle plugin does that](https://github.com/grblHAL/Plugins_spindle/blob/53d72fae862bd45b447e674186c0165ccf4b9a5b/huanyang.c#L361-L368) to control a VFD.

> I am sure you are not in a position to answer practical aspects of it but just a wild idea from me for your comments please.

I am not sure how the current spindle sync code will work or if it will need modification. There is a large change in spindle speed (and thus feed rate) when the spindle changes direction...

**Comment by Team-Deepflights** [comment]
I bought another encoder with 200 PPM and Z-index, this works perfect. Somehow too perfect, my RPM raise and fall about 5-7 RPM, I suppose this is normal?
When all hardware is finished I dig in into the PID tuning.
Very nice feature, thx a lot for that :-)
